### ISSUE TITLE
One way binding

### ISSUE THREAD

**Issue by samselikoff** [miguelcobain/ember-cli-selectize#53]
We're having some trouble when binding value to a query param, with null/undefined. Is there any way to disable two-way binding? Ideally the change would only send out an action, rather than updating the `value` that was passed in.

**Comment by miguelcobain** [comment]
Do you mean `selection` instead of `value`?

**Comment by samselikoff** [comment]
Either, really just want the select box to update if the value passed into it updates, but when user changes select box, just want it to send out an action (and not update either value or selection)

**Comment by miguelcobain** [comment]
That sounds exactly like the definition of "data down, actions up".
Didn't think of that use case.

We may need to do some slight changes to make this conform that paradigm.

**Comment by samselikoff** [comment]
right. Maybe an opt-in to disable the two-way binding so it's backwards compatible.

**Comment by miguelcobain** [comment]
https://github.com/miguelcobain/ember-cli-selectize/blob/master/addon%2Fcomponents%2Fember-selectize.js#L256

I'd say that, in the `_updateSelection` function, we would either trigger an action or update the value, but not both?

**Comment by samselikoff** [comment]
yah, again you can leave both on by default if you dont want to break bwds compatibility, but if `twoWay=false` don't run line 257

**Comment by samselikoff** [comment]
it's going to be a bit more than that line. I commented it out locally, but there's still some wackiness going on

**Comment by miguelcobain** [comment]
Is your select multiple or single?

**Comment by samselikoff** [comment]
single. I think it has to do with `_valueDidChange`, but commenting out that method breaks the "data down" part (select doesn't represent value passed in)

**Comment by miguelcobain** [comment]
There are a couple of  `set('selection',...)`. Any of those may be the culprit.

**Comment by samselikoff** [comment]
right. So it looks like I actually got what I want with this:

```js
// components/ember-selectize.js
import Component from 'ember-cli-selectize/components/ember-selectize';

Component.reopen({
  _updateSelection: function(selection) {
    // allow the observers and computed properties to run first
    Ember.run.schedule('actions', this, function() {
      var value = this.get('value');
      this.sendAction('select-item', selection, value);
    });
  },
  _valueDidChange: Ember.observer('value', function() {
  }),
});

export default Component;
```

This lets me pass in `selection`, but `ember-selectize` itself doesn't call .set, it just sends out the action.

This didn't work when I just passed in `value` though.

**Comment by miguelcobain** [comment]
Glad you made it work for now.

I'll restructure the code a bit more so that we have a single point of updates.
That way we can have a better control.

Meanwhile, let's keep this open.
Thanks.

**Comment by knownasilya** [comment]
Would love a `twoWay=false` option.

**Comment by miguelcobain** [comment]
I would prefer to infer that by checking if actions are bound or not.
Basically, if `select-item` action is used, don't update `selection` or `value`.

This would of course break if people are using both actions and two-way bindings. But I kind of think of that as an anti-pattern.

Thoughts?

**Comment by knownasilya** [comment]
Agreed. I'm assuming this would work for `add-item` and `remove-item` as well.

**Comment by miguelcobain** [comment]
From [all the actions](https://github.com/miguelcobain/ember-cli-selectize#actions), maybe only if any of`select-item`, `add-item` or `remove-item` is set, we don't update `selection` or `value`.

`create-item` and `update-filter` shouldn't trigger this behavior because people who wish to use two bindings may need them.

**Comment by samselikoff** [comment]
@knownasilya I've also started wrapping selectize (and other libs) in my own component code and using variables local to that wrapper for the two-way binding, but only sending actions out. This is a short-term way to enforce one-way binding in your app. (Also, I think wrapping 3rd-party components in your application is generally a good pattern).

**Comment by knownasilya** [comment]
@samselikoff feel like sharing your wrapper?

**Comment by knownasilya** [comment]
@miguelcobain were you able to make those changes for the actions?

**Comment by samselikoff** [comment]
I'm not sure what you mean by final value.

I usually just have something like `{{ted-select content=content selection=selection}}` and so on, and internally to the component I'd have

```js
export default Ember.Component.extend({
  _selection: Ember.computed.oneWay('selection');

  emitActions: Ember.observer('_selection', function() {
    if (this.get('_selection') !== this.get('selection') {
      this.sendAction('on-select', this.get('_selection');
    }
  }
});
```

and then the component's template would be something like

```hbs
{{view 'select' content=content selection=_selection}}
```

**Comment by knownasilya** [comment]
So input is `[{ id: 1, value: 'test' }]` and output is `[1]`, which would be the final output, since ember-selectize doesn't support `value` and `selection` needs to be transformed.

**Comment by samselikoff** [comment]
output meaning, the parameter of the `on-select` action?

**Comment by knownasilya** [comment]
See https://gist.github.com/knownasilya/f200d462e6c3443f41ac, trying to get the selection to turn into and object, but having a hard time. It works on initial load if selection is an array of ids (internally it's replaced with object versions), but on-select it reverts back to the id version. I've tried a few things but keep running into stack overflows.. I can't wait for `(mut)`.

**Comment by samselikoff** [comment]
did you try `_selection: Ember.computed.oneWay('selection');`? If you do that selectize should never mutate the thing you pass in

**Comment by knownasilya** [comment]
Is there a way to do `oneWay` with a full-blown computed property? Since I need to transform the users data from array of values, to an array of objects.

**Comment by samselikoff** [comment]
I dont think so, can't you just make sure your CP is only composed of other readonly properties?

**Comment by Fenntasy** [comment]
> From all the actions, maybe only if any of select-item, add-item or remove-item is set, we don't update selection or value.

@miguelcobain is it true in the current version?

**Comment by knownasilya** [comment]
Most people are using ember-power-select now a days.

**Comment by Fenntasy** [comment]
@knownasilya thanks for the tip
